core/src/main/scala/org/broadinstitute/dsde/rawls/entities/compact/CompactEntityProvider.scala

```diff
@@ -66,7 +73,50 @@ class CompactEntityProvider(requestArguments: EntityRequestArguments, repository
   override def batchUpsertEntities(
     entityUpdates: Source[EntityUpdateDefinition, _],
     parentContext: RawlsRequestContext
-  ): Future[Int] = ???
+  ): Future[Int] = {
+
+    // Translate the input stream entityUpdates to a stream of Entity by applying the updates to
+    // a pre-existing entity (for updates) or a blank entity (for inserts)
+    val entitySource: Source[Entity, _] = entityUpdates.map { updateDefinition =>
+      // TODO CORE-428: for updates, look ahead (some quantity) in the update definitions and
+      //   retrieve the existing entities from the db
+      // for inserts, start with an empty entity
+      val baseEntity = Entity(updateDefinition.name, updateDefinition.entityType, Map())
+      // update the starting entity with the user's operations
+      applyOperationsToEntity(baseEntity, updateDefinition.operations)
+    }
+
+    // Group the entities-to-be-saved into batches to optimize our SQL interactions
+    val batches: Source[Seq[Entity], _] = entitySource.groupedWeighted(config.maxSqlBatchSizeBytes)(calculateEntitySize)
+
+    // For each batch, generate the db action to write it to the database
+    val batchActionsSource: Source[ReadWriteAction[Int], _] = batches
+      .map { batch =>
+        logger.info(s"batch upsert: batch of ${batch.size} entities")
+        insertBatch(batch)
+      }
+
+    // Materialize the batch actions into a sequence and convert to a single DBIO action
+    val batchActionsF: Future[ReadWriteAction[Int]] = batchActionsSource
+      .runWith(Sink.seq)
+      .map(DBIO.sequence(_).map(_.sum))
+
+    // Convert the Future[ReadWriteAction] to a Source[Int] by executing the db actions.
+    // Note that since all the ReadWriteActions are fused via DBIO.sequence above,
+    // there is only one action to execute inside a transaction.
+    val dbResultsSource: Source[Int, _] = Source.futureSource(batchActionsF.map { dbAction =>
+      Source.future(repository.dataSource.inTransaction(_ => dbAction))
+    })
+
+    // Finally, run the Source. Ignore the results from the db; they are not necessary.
+    val dbResults: Future[Int] = dbResultsSource.runWith(Sink.head)
```

Comment:
Slightly confused by this comment; it seems like we're returning the results from the db instead of ignoring them?

Reply:
thank you for catching that; fixed. The comment was out of date.